src/server/actions/s3.actions.ts

```diff
@@ -0,0 +1,33 @@
+'use server'
+
+import { getS3Client, s3BucketName } from '@/server/aws'
+import { MinimalJobInfo, minimalJobInfoSchema, minimalStudyInfoSchema, StudyDocumentType } from '@/lib/types'
+import { pathForStudyDocumentFile, pathForStudyJobCode } from '@/lib/paths'
+import { createPresignedPost } from '@aws-sdk/s3-presigned-post'
+import { userAction, z } from '@/server/actions/wrappers'
+
+export const signedUrlForCodeUploadAction = userAction(async (jobInfo: MinimalJobInfo) => {
+    const prefix = pathForStudyJobCode(jobInfo)
+
+    return await createPresignedPost(getS3Client(), {
+        Bucket: s3BucketName(),
+        Conditions: [['starts-with', '$key', prefix]],
+        Expires: 3600,
+        Key: prefix + '/${filename}', // single quotes are intentional, S3 will replace ${filename} with the filename
+    })
+}, minimalJobInfoSchema)
+
+export const signedUrlForStudyFileUploadAction = userAction(
+    async ({ studyInfo, documentType, filename }) => {
+        return await createPresignedPost(getS3Client(), {
+            Bucket: s3BucketName(),
+            Expires: 3600,
+            Key: pathForStudyDocumentFile(studyInfo, documentType, filename),
+        })
+    },
+    z.object({
+        studyInfo: minimalStudyInfoSchema,
+        documentType: z.nativeEnum(StudyDocumentType),
+        filename: z.string(),
+    }),
```

Comment:
see my comment above about what `key` does when adding to the Form for uploading. If we can control the path, then we only need a single signed url generator and each file can set it's own path under the study